File: GitToolBox/src/main/java/zielu/gittoolbox/util/DisposeSafeCallable.java
package zielu.gittoolbox.util;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import java.util.concurrent.Callable;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.jetbrains.annotations.NotNull;

public class DisposeSafeCallable<T> implements Callable<T> {
  private final Logger log = Logger.getInstance(getClass());
  private final Project project;
  @NotNull
  private final Callable<T> operation;
  private final T disposedResult;

  public DisposeSafeCallable(@NotNull Project project, @NotNull Callable<T> operation, T disposedResult) {
    this.project = project;
    this.operation = operation;
    this.disposedResult = disposedResult;
  }

  @Override
  public T call() throws Exception {
    try {
      return operation.call();
    } catch (AssertionError error) {
      if (project.isDisposed()) {
        log.debug("Project already disposed", error);
        return disposedResult;
      } else {
        log.error(error);
        return disposedResult;
      }
    }
  }

  @Override
  public String toString() {
    return new ToStringBuilder(this)
        .append("operation", operation)
        .append("disposedResult", disposedResult)
        .toString();
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/util/DisposeSafeCallable.kt
package zielu.gittoolbox.util

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import org.apache.commons.lang3.builder.ToStringBuilder
import org.apache.commons.lang3.builder.ToStringStyle
import java.util.concurrent.Callable

internal class DisposeSafeCallable<T> (
  private val project: Project,
  private val operation: Callable<T>,
  private val disposedResult: T
) : Callable<T> {

  @Throws(Exception::class)
  override fun call(): T {
    return try {
      if (project.isDisposed) {
        disposedResult
      } else {
        operation.call()
      }
    } catch (error: AssertionError) {
      if (project.isDisposed) {
        log.debug("Project already disposed", error)
        disposedResult
      } else {
        log.error(error)
        disposedResult
      }
    }
  }

  override fun toString(): String {
    return ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE)
      .append("operation", operation)
      .append("disposedResult", disposedResult)
      .toString()
  }

  private companion object {
    private val log = Logger.getInstance(DisposeSafeCallable::class.java)
  }
}
